# Exercise 1

#### (a)

When looking at all the juncture points and finding out how many each one is connected to and whether they are connected to the top or bottom we find the system of equations is as follows:

   $ 3V_1 - V_2 - V_3 - V_4 = 5$<br>
   $-V_1 + 4V_2 - V_3 - V_4 = 5$<br>
   $-V_1 - V_2 + 4V_3 - V_4 = 0$<br>
   $-V_1 - V_2 - V_3 + 3V_4 = 0$<br>
   
#### (b)

From here it is easy to set up a system of linear equations in matrix form Av=b where v is the unknown voltages:

         [[3 -1 -1 0 | 5]
         [-1 4 -1 -1 | 5]
         [-1 -1 4 -1 | 0]
         [0 -1 -1 3 | 0]]

In [1]:
#Kaleb Kautzsch
#(c)
import scipy as sp
import scipy.linalg as la

a = sp.array([[3,-1,-1,0],[-1,4,-1,-1],[-1,-1,4,-1],[0,-1,-1,3]]) #The matrix as previously stated
b = sp.array([[5],[5],[0],[0]]) #We have to divide by 2 for the intial voltage input because it is connected to two resistors
v = la.solve(a,b)

print(v, "The voltages")

[[3.33333333]
 [3.        ]
 [2.        ]
 [1.66666667]] The voltages


# Exercise 2

In [27]:
#Kaleb Kautzsch
import numpy as np
import scipy.linalg as la

N = 10
a = np.zeros([N,N])
b = np.zeros([N,1])
b[0,0] = 5
b[1,0] = 5

def matrix():
    for i in range(N):
        for j in range(N):
            if abs(i-j) == 1 or abs(i-j) == 2:
                a[i][j] = -1
            elif i == j:
                a[i][j] = 4
    a[0,0] = 3
    a[N-1,N-1] = 3
    return a

matrix()
v = la.solve(matrix(),b)
print(matrix(),' Matrix A\n\n',v,' Solution v')

[[ 3. -1. -1.  0.  0.  0.  0.  0.  0.  0.]
 [-1.  4. -1. -1.  0.  0.  0.  0.  0.  0.]
 [-1. -1.  4. -1. -1.  0.  0.  0.  0.  0.]
 [ 0. -1. -1.  4. -1. -1.  0.  0.  0.  0.]
 [ 0.  0. -1. -1.  4. -1. -1.  0.  0.  0.]
 [ 0.  0.  0. -1. -1.  4. -1. -1.  0.  0.]
 [ 0.  0.  0.  0. -1. -1.  4. -1. -1.  0.]
 [ 0.  0.  0.  0.  0. -1. -1.  4. -1. -1.]
 [ 0.  0.  0.  0.  0.  0. -1. -1.  4. -1.]
 [ 0.  0.  0.  0.  0.  0.  0. -1. -1.  3.]]  Matrix A

 [[4.12573674]
 [3.9194499 ]
 [3.45776031]
 [3.09430255]
 [2.69155206]
 [2.30844794]
 [1.90569745]
 [1.54223969]
 [1.0805501 ]
 [0.87426326]]  Solution v


# Results for N = 4 and N = 10
### When N = 4:

                                    A = [[ 3. -1. -1.  0.]
                                         [-1.  4. -1. -1.]
                                         [-1. -1.  4. -1.]
                                         [ 0. -1. -1.  3.]]
                                         
                                    v = [[3.33333333]
                                         [3.        ]
                                         [2.        ]
                                         [1.66666667]]
                                         
This value for "v" is the same calculated in my first code.

### When N = 10:

                                    A = [[ 3. -1. -1.  0.  0.  0.  0.  0.  0.  0.]
                                         [-1.  4. -1. -1.  0.  0.  0.  0.  0.  0.]
                                         [-1. -1.  4. -1. -1.  0.  0.  0.  0.  0.]
                                         [ 0. -1. -1.  4. -1. -1.  0.  0.  0.  0.]
                                         [ 0.  0. -1. -1.  4. -1. -1.  0.  0.  0.]
                                         [ 0.  0.  0. -1. -1.  4. -1. -1.  0.  0.]
                                         [ 0.  0.  0.  0. -1. -1.  4. -1. -1.  0.]
                                         [ 0.  0.  0.  0.  0. -1. -1.  4. -1. -1.]
                                         [ 0.  0.  0.  0.  0.  0. -1. -1.  4. -1.]
                                         [ 0.  0.  0.  0.  0.  0.  0. -1. -1.  3.]]
                                         
                                     v = [[4.12573674]
                                          [3.9194499 ]
                                          [3.45776031]
                                          [3.09430255]
                                          [2.69155206]
                                          [2.30844794]
                                          [1.90569745]
                                          [1.54223969]
                                          [1.0805501 ]
                                          [0.87426326]]